cartographer_ros/cartographer_ros/trajectory_options.cc

```diff
@@ -69,6 +72,29 @@ TrajectoryOptions CreateTrajectoryOptions(
   return options;
 }
 
+TrajectoryOptions CreateTrajectoryOptions(
+    ::cartographer::common::LuaParameterDictionary* lua_parameter_dictionary,
+    ::cartographer::common::LuaParameterDictionary* initial_trajectory_pose) {
+  TrajectoryOptions options = CreateTrajectoryOptions(lua_parameter_dictionary);
+  *options.trajectory_builder_options.mutable_initial_trajectory_pose() =
+      CreateInitialTrajectoryPose(initial_trajectory_pose);
+  return options;
+}
+
+::cartographer::mapping::proto::InitialTrajectoryPose
+CreateInitialTrajectoryPose(
+    ::cartographer::common::LuaParameterDictionary* lua_parameter_dictionary) {
+  ::cartographer::mapping::proto::InitialTrajectoryPose pose;
+  pose.set_to_trajectory_id(
+      lua_parameter_dictionary->GetNonNegativeInt("to_trajectory_id"));
+  *pose.mutable_relative_pose() =
+      cartographer::transform::ToProto(cartographer::transform::FromDictionary(
+          lua_parameter_dictionary->GetDictionary("relative_pose").get()));
+  pose.set_timestamp(
+      cartographer::common::ToUniversal(FromRos(ros::Time::now())));
```

Comment:
You want to read the timestamp from the command line. The timestamp does not specify when the new trajectory started, rather it indicates at what time you you want to take the global pose from the to_trajectory apply your relative pose and make that your starting position.

Reply:
Yes but when you want to use it is so that not simple. I think that starting a new trajectory with initial pose does not play well with start_trajectory_main because you must have access to the pose of the trajectory at a certain time. Maybe I am wrong but for me start_trajectory is used "interactively" so it means that:
- you have cartographer_ros running
- you see the construction of the map and at particular time you decide to start a new trajectory
In this case providing a timestamp is cumbersome, you have to find the current time and use it to start the new trajectory with an identity matrix.

To support what @cschuet said and the use case above I think that the timestamp should be optional and if not provided set to current ros time.

Reply:
Ok, I see that use case makes sense to me. Also I am not super familiar with ROS. I think my concern was that taking the current time may lead to unexpected behavior when the relation between recorded timestamps and system time is not what you expect. But reading up on `ros::Time::now()` it seems that within a rosbag playback it returns the current time *in the simulation*. If that's the case I am ok with having it default to `ros::Time::now()` when no timestamp is provided.

Reply:
Yes, `ros::Time::now()` was to compatible with ros simulated time from rosbag.

Thank you guys for review. I will update the PR based on your comment.